feat(stop_filter): add unit tests to stop filter node

interimadd opened the pull request:
## Description

This PR extracts the stop filter logic from the stop filter node code and adds unit tests for both.
The motivation for this pull request is to improve testability and portability by separating the pure logic from the code that depends on ROS2 node functionality.

## How was this PR tested?

### Unit test result

Confirmed that all 7 unit tests passed.

```
~/autoware$ cd ~/autoware && colcon build --symlink-install --cmake-args -DBUILD_TESTING=true --packages-select autoware_stop_filter && source install/setup.bash && colcon test --packages-select autoware_stop_filter --event-handlers console_direct+

1: [==========] Running 7 tests from 3 test suites.
1: [----------] Global test environment set-up.
1: [----------] 4 tests from StopFilterTest
1: [ RUN      ] StopFilterTest.FilterStopWhenStopped
1: [       OK ] StopFilterTest.FilterStopWhenStopped (0 ms)
1: [ RUN      ] StopFilterTest.FilterStopWhenNotStopped
1: [       OK ] StopFilterTest.FilterStopWhenNotStopped (0 ms)
1: [ RUN      ] StopFilterTest.FilterStopOnlyLinearVelocityBelowThreshold
1: [       OK ] StopFilterTest.FilterStopOnlyLinearVelocityBelowThreshold (0 ms)
1: [ RUN      ] StopFilterTest.FilterStopOnlyAngularVelocityBelowThreshold
1: [       OK ] StopFilterTest.FilterStopOnlyAngularVelocityBelowThreshold (0 ms)
1: [----------] 4 tests from StopFilterTest (0 ms total)
1: 
1: [----------] 2 tests from StopFilterProcessorTest
1: [ RUN      ] StopFilterProcessorTest.TestCreateStopFlagMsgMoving
1: [       OK ] StopFilterProcessorTest.TestCreateStopFlagMsgMoving (0 ms)
1: [ RUN      ] StopFilterProcessorTest.TestCreateFilteredMsgStopped
1: [       OK ] StopFilterProcessorTest.TestCreateFilteredMsgStopped (0 ms)
1: [----------] 2 tests from StopFilterProcessorTest (0 ms total)
1: 
1: [----------] 1 test from StopFilterNodeTest
1: [ RUN      ] StopFilterNodeTest.TestStopDetection
1: [       OK ] StopFilterNodeTest.TestStopDetection (139 ms)
1: [----------] 1 test from StopFilterNodeTest (139 ms total)
1: 
1: [----------] Global test environment tear-down
1: [==========] 7 tests from 3 test suites ran. (139 ms total)
1: [  PASSED  ] 7 tests.
```

### Code coverage

Confirmed that the code line coverage is almost 100%.

<img width="1024" height="245" alt="image" src="https://github.com/user-attachments/assets/6102fc0b-2ddb-4e20-ac72-df97f4c6c010" />

### Launch as node

Confirmed that stop_filter_node can be launched as a ROS2 node.

```sh
~/autoware$ source install/local_setup.bash
~/autoware$ ros2 run autoware_stop_filter autoware_stop_filter_node --ros-args -p wz_threshold:=0.5 -p vx_threshold:=0.5
```

```sh
~/autoware$ ros2 node info /stop_filter
/stop_filter
  Subscribers:
    /input/odom: nav_msgs/msg/Odometry
    /parameter_events: rcl_interfaces/msg/ParameterEvent
  Publishers:
    /debug/stop_flag: autoware_internal_debug_msgs/msg/BoolStamped
    /output/odom: nav_msgs/msg/Odometry
    /parameter_events: rcl_interfaces/msg/ParameterEvent
    /rosout: rcl_interfaces/msg/Log
  Service Servers:
    /stop_filter/describe_parameters: rcl_interfaces/srv/DescribeParameters
    /stop_filter/get_parameter_types: rcl_interfaces/srv/GetParameterTypes
    /stop_filter/get_parameters: rcl_interfaces/srv/GetParameters
    /stop_filter/list_parameters: rcl_interfaces/srv/ListParameters
    /stop_filter/set_parameters: rcl_interfaces/srv/SetParameters
    /stop_filter/set_parameters_atomically: rcl_interfaces/srv/SetParametersAtomically
```

## Notes for reviewers

The integration test named `TestStopDetection` is disabled by default because it takes a long time to run.

## Interface changes

None.

## Effects on system behavior

None.

sasakisasaki commented on localization/autoware_stop_filter/include/autoware/stop_filter/stop_filter_node.hpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 TierIV
+// Copyright 2025 TierIV
```
Perhaps it looks better to keep `2021` to show from when the code is being licensed :+1: 
```suggestion
// Copyright 2021 TierIV
```

sasakisasaki commented on localization/autoware_stop_filter/src/stop_filter.cpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 TierIV
+// Copyright 2025 TierIV
```
Same as above
```suggestion
// Copyright 2021 TierIV
```

sasakisasaki commented on localization/autoware_stop_filter/src/stop_filter_node.cpp:
```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 TierIV
```
As this is newly added file, let's use `2025`
```suggestion
// Copyright 2025 TIER IV
```

sasakisasaki commented on localization/autoware_stop_filter/test/test_stop_filter.cpp:
```diff
@@ -0,0 +1,82 @@
+// Copyright 2025 TierIV
```
```suggestion
// Copyright 2025 TIER IV
```

sasakisasaki reviewed (approved):
Great work!!! :100: 

Although I only had a few minor comments, I support merging this PR for the following reasons:
- Previously, many lines were difficult to test because they were inside callback functions that require the node to be spun.
- This PR refactors those sections into smaller, more modular chunks, which makes them easier to test independently—without spinning the node (i.e., without triggering callbacks).
- The changes also improve both readability and testability of the code.

interimadd commented:
@sasakisasaki 
Thanks for the review! I've updated the copyright.
I'm glad you like the idea of creating a class that's testable without launch a node.

sasakisasaki reviewed (changes requested):
@YamatoAndo @isamu-takagi Thank you very much for your review!

@interimadd It seems the cppcheck is failing after the fix. Could you kindly fix so the CI passes? Thank you in advance.
https://github.com/autowarefoundation/autoware_core/actions/runs/16365945094/job/46243808205?pr=573

interimadd commented:
Thank you for all the reviews! And sorry for overlooking some guidelines.
I have fixed following items.
- Fixed cppcheck issues
- Moved the .hpp file from `autoware/stop_filter/include` to `src`
- Modified CMakeLists.txt to use the recommended macro

Could you check it again?

YamatoAndo commented on localization/autoware_stop_filter/src/stop_filter.cpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 TierIV
+// Copyright 2025 TIER IV
```
keep `2021`
```
// Copyright 2021 TIER IV
```

or 
```
// Copyright 2021-2025 TIER IV
```

YamatoAndo commented on localization/autoware_stop_filter/src/stop_filter.hpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 TierIV
+// Copyright 2025 TIER IV
```
keep 2021

// Copyright 2021 TIER IV
or

// Copyright 2021-2025 TIER IV

interimadd commented on localization/autoware_stop_filter/CMakeLists.txt:
```diff
@@ -6,15 +6,31 @@ autoware_package()
-rclcpp_components_register_node(${PROJECT_NAME}
-  PLUGIN "autoware::stop_filter::StopFilter"
+ament_auto_add_library(${PROJECT_NAME}_ros SHARED
+  src/stop_filter_node.cpp
+)
```
The node name remains `autoware_stop_filter_node`, so it can be launched with the following command:

```sh
ros2 run autoware_stop_filter autoware_stop_filter_node --ros-args -p wz_threshold:=0.5 -p vx_threshold:=0.5
```

If the library were also named `autoware_stop_filter_node`, it would conflict with the executable's name, causing the build error shown below.
```sh
~/autoware$ cd ~/autoware && colcon build --symlink-install --cmake-args -DBUILD_TESTING=true --packages-select autoware_stop_filter

--- stderr: autoware_stop_filter                         
CMake Error at /opt/ros/humble/share/rclcpp_components/cmake/rclcpp_components_register_node.cmake:72 (add_executable):
  add_executable cannot create target "autoware_stop_filter_node" because
  another target with the same name already exists.  The existing target is a
  shared library created in source directory
  "/home/takahisaishikawa/autoware/src/core/autoware_core/localization/autoware_stop_filter".
  See documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  CMakeLists.txt:19 (rclcpp_components_register_node)
```

The target library named `autoware_stop_filter_ros` is required to run unit tests and to create executable.
So, I would like to keep the current statement.

sasakisasaki reviewed (approved):
I re-checked the change. My approval reason is same as [this comment](https://github.com/autowarefoundation/autoware_core/pull/573#pullrequestreview-3028777341)
